### PR TITLE
feat: add display_callback option

### PR DESCRIPTION
**Pull request by theHamsta**
Addresses #47, #48

@przepompownia does this solve your problems? If you could install a listener for `dap.listeners.after.variables` to decide which variables you want to evaluate. Store the results of  the evaluations somewhere and `refresh` of this plugin once you got the response for all your evaluate requests.

### CODE REVIEW

File: lua/nvim-dap-virtual-text.lua [hunk 1]
```diff
@@ -46,6 +46,15 @@ local options = {
   -- filter references (not definitions) pattern when `all_references` is activated
   -- (Lua gmatch pattern, default filters out Python modules)
   filter_references_pattern = '<module',
+  --- A callback that determines how a variable is displayed or whether it should be omitted
+  --- @param variable table https://microsoft.github.io/debug-adapter-protocol/specification#Types_Variable
+  --- @param buf number
+  --- @param stackframe table https://microsoft.github.io/debug-adapter-protocol/specification#Types_StackFrame
+  --- @return string|nil A text how the virtual text should be displayed or nil, if this variable shouldn't be displayed
+  --- @diagnostic disable-next-line: unused-local
+  display_callback = function(variable, buf, stackframe)
```
Comment (theHamsta) [inline lua/nvim-dap-virtual-text.lua:55]: you think that the `range` (line,col start, line, col end) of the variable is relevant?

Comment (theHamsta) [reply lua/nvim-dap-virtual-text.lua]: added the whole tree-sitter node

**Comment by theHamsta** [comment]
Modifying `value` of a variable would also change how it is displayed in `dapui` too. Maybe this is more what you want.

**Comment by przepompownia** [comment]
@theHamsta thank you for spending time for this change.

Ath the moment I at least checked if some simple filter works:
```lua
local api = vim.api

---@param variable any
---@param buf any
---@return boolean
local function skipDisplayingVariable(variable, buf)
  local filetype = vim.bo[buf].filetype
  if variable.name == '$this' and filetype == 'php' then
    return true
  end
  return false
end

require('nvim-dap-virtual-text').setup {
  enabled = true,
  enabled_commands = true,
  highlight_changed_variables = true,
  highlight_new_as_changed = false,
  show_stop_reason = true,
  commented = false,
  only_first_definition = false,
  all_references = true,
  filter_references_pattern = '<module',
  virt_text_pos = 'eol',
  all_frames = false,
  virt_lines = false,
  virt_text_win_col = nil,
  display_callback = function(variable, bufId, _stackframe, _node)
    if true == skipDisplayingVariable(variable, bufId) then
      return ''
    end
    return variable.name .. ': ' .. variable.value
  end,
}
```

image

The whole example (prepared to test DAP evaluations) is available on https://github.com/przepompownia/dap-ui-test (after `make start` and `make virtual-text-callback`). 

I need to resolve some neodev+sumneko issues before checking how it works with evaluation. Do you suggest storing results from the `after.variables` listener for a performance reasons?

**Comment by theHamsta** [comment]
> @theHamsta thank you for spending time for this change.
> 
> Ath the moment I at least checked if some simple filter works:
> 
> ```lua
> local api = vim.api
> 
> ---@param variable any
> ---@param buf any
> ---@return boolean
> local function skipDisplayingVariable(variable, buf)
>   local filetype = vim.bo[buf].filetype
>   if variable.name == '$this' and filetype == 'php' then
>     return true
>   end
>   return false
> end
> 
> require('nvim-dap-virtual-text').setup {
>   enabled = true,
>   enabled_commands = true,
>   highlight_changed_variables = true,
>   highlight_new_as_changed = false,
>   show_stop_reason = true,
>   commented = false,
>   only_first_definition = false,
>   all_references = true,
>   filter_references_pattern = '<module',
>   virt_text_pos = 'eol',
>   all_frames = false,
>   virt_lines = false,
>   virt_text_win_col = nil,
>   display_callback = function(variable, bufId, _stackframe, _node)
>     if true == skipDisplayingVariable(variable, bufId) then
>       return ''
>     end
>     return variable.name .. ': ' .. variable.value
>   end,
> }
> ```
> 
> image
> 
> The whole example (prepared to test DAP evaluations) is available on https://github.com/przepompownia/dap-ui-test (after `make start` and `make virtual-text-callback`).
> 
> I need to resolve some neodev+sumneko issues before checking how it works with evaluation. Do you suggest storing results from the `after.variables` listener for a performance reasons?

I made upstream nvim-dap store them in the stackframe. See the type annotations that I added in this PR. Each stack frame has a `scopes` field that has `variables`

**Pull request merged**